src/lucky/param_serializable.cr

```diff
@@ -0,0 +1,177 @@
+            end
+          {% else %}
+            %val = if %param_key_value
+              %data = params.nested{% if metadata[:is_file] %}_file{% end %}?(%param_key_value.not_nil!)
+
+              {% if metadata[:is_serializable] %}
+              Lucky::Params.from_hash(%data)
+              {% else %}
+              %data.get(:{{ metadata[:type_declaration].var.id }})
+              {% end %}
+            else
+              params.get{% if metadata[:is_file] %}_file{% end %}?(:{{ metadata[:type_declaration].var.id }})
+            end
+          {% end %}
+
+          if %val.nil?{% if metadata[:nilable] %} || %val == ""{% end %}
+            default_or_nil = {{ metadata[:type_declaration].value ? metadata[:type_declaration].value : nil }}
+            {% if metadata[:nilable] %}
+              if %val.nil? && default_or_nil.nil?
+                @{{ metadata[:type_declaration].var.id }} = nil
+              else
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: default_or_nil, param_key: %param_key_value)
+              end
+            {% else %}
+              if default_or_nil.nil?
+                raise Lucky::MissingParamError.new <<-ERROR
+                {{ @type }} is missing value for required param "{{ metadata[:type_declaration] }}"
+                ERROR
+              else
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: default_or_nil, param_key: %param_key_value)
+              end
+            {% end %}
+          else
+            {% if metadata[:is_serializable] %}
+            @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: {{ metadata[:core_type] }}.from_params(%val.as(Lucky::Params)), param_key: %param_key_value)
+            {% else %}
+              # NOTE: these come from Avram directly
+              %result = {{ metadata[:core_type] }}::Lucky.parse(%val)
+
+              if %result.is_a? Avram::Type::SuccessfulCast
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: %result.value.as({{ metadata[:type] }}), param_key: %param_key_value)
+              else
+                raise Lucky::InvalidParamError.new(
+                  param_name: "{{ metadata[:type_declaration].var.id }}",
+                  param_value: %val.to_s,
+                  param_type: "{{ metadata[:type] }}"
+                )
+              end
+            {% end %}
+          end
+        {% end %}
+      {% end %}
+    end
+
+    macro param(type_declaration, param_key = nil)
```

Comment:
Decided to go with a custom macro here because as the user, you specify `name : String`, but it's really `name : Lucky::PermittedParam(String)`. This also gives us the ability to determine if the value is `nil` by intention, or nil because no value was passed in.